lib/charms/operator_libs_linux/v0/grub.py

```diff
@@ -29,4 +37,259 @@
 # to 0 if you are raising the major API version
 LIBPATCH = 1
 
-# TODO: add your code here! Happy coding!
+LIB_CONFIG_DIRECTORY = Path("/var/lib/charm-grub/")
+GRUB_CONFIG = Path("/etc/default/grub.d/") / "95-juju-charm.cfg"
+CONFIG_HEADER = f"""
+# This config file was produced by grub lib v{LIBAPI}.{LIBPATCH}.
+# https://charmhub.io/operator-libs-linux/libraries/grub
+"""
+FILE_LINE_IN_DESCRIPTION = "#   {path}"
+CONFIG_DESCRIPTION = """
+# This file represents the output of the grub lib, which can combine multiple
+# configurations into a single file like this.
+#
+# Original files:
+#  {configs}
+#
+# If you change this file, run 'update-grub' afterwards to update
+# /boot/grub/grub.cfg.
+# For full documentation of the options in this file, see:
+#   info -f grub -n 'Simple configuration'
+"""
+
+# TODO: add better description to lib
+
+
+class ValidationError(ValueError):
+    """Exception representing value validation error."""
+
+    def __init__(self, key: str, message: str) -> None:
+        self._key = key
+        self._message = message
+        super().__init__(message)
+
+    @property
+    def key(self) -> str:
+        """Get name of key, which cause this error."""
+        return self._key
+
+    @property
+    def message(self) -> str:
+        """Get error message."""
+        return self._message
+
+
+class IsContainerError(Exception):
+    """Exception if local machine is container."""
+
+
+class ApplyError(Exception):
+    """Exception if applying new config failed."""
+
+
+def _parse_config(data: str) -> Dict[str, str]:
+    """Parse config file lines.
+
+    This function is capable to update single key value if it's used with $ symbol.
```

Comment:
Second line of docstring might sound better like this:

`This function is capable of updating single key values using the "$" symbol`

Reply:
Can you check it again, I add full docstring with example.